sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala

```diff
@@ -78,37 +80,52 @@ object ResolveTableValuedFunctions extends Rule[LogicalPlan] {
   private val builtinFunctions: Map[String, TVF] = Map(
     "range" -> Map(
       /* range(end) */
-      tvf("end" -> LongType) { case Seq(end: Long) =>
-        Range(0, end, 1, None)
+      tvf("end" -> LongType) { case (tvf, args @ Seq(end: Long)) =>
+        validateInputDimension(tvf, 1)
+        Range(0, end, 1, None, tvf.outputNames.headOption)
       },
 
       /* range(start, end) */
-      tvf("start" -> LongType, "end" -> LongType) { case Seq(start: Long, end: Long) =>
-        Range(start, end, 1, None)
+      tvf("start" -> LongType, "end" -> LongType) {
+        case (tvf, args @ Seq(start: Long, end: Long)) =>
+          validateInputDimension(tvf, 1)
+          Range(start, end, 1, None, tvf.outputNames.headOption)
       },
 
       /* range(start, end, step) */
       tvf("start" -> LongType, "end" -> LongType, "step" -> LongType) {
-        case Seq(start: Long, end: Long, step: Long) =>
-          Range(start, end, step, None)
+        case (tvf, args @ Seq(start: Long, end: Long, step: Long)) =>
+          validateInputDimension(tvf, 1)
+          Range(start, end, step, None, tvf.outputNames.headOption)
       },
 
       /* range(start, end, step, numPartitions) */
       tvf("start" -> LongType, "end" -> LongType, "step" -> LongType,
           "numPartitions" -> IntegerType) {
-        case Seq(start: Long, end: Long, step: Long, numPartitions: Int) =>
-          Range(start, end, step, Some(numPartitions))
+        case (tvf, args @ Seq(start: Long, end: Long, step: Long, numPartitions: Int)) =>
+          validateInputDimension(tvf, 1)
+          Range(start, end, step, Some(numPartitions), tvf.outputNames.headOption)
       })
   )
 
+  private def validateInputDimension(tvf: UnresolvedTableValuedFunction, expectedNumCols: Int)
```

Comment:
Conceptually, `expectedNumCols` needs to be part of `type/class TVF`. This is the num of TVF's output arguments.